src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignedCmsTests.cs

```diff
@@ -890,5 +890,488 @@ public static void SignWithImplicitSubjectKeyIdentifier()
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            using (X509Certificate2 cert = Certificates.ECDsaP256Win.GetCertificate())
+            using (ECDsa key = ECDsa.Create())
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                Assert.Throws<CryptographicException>(() => cms.ComputeSignature(signer));
+            }
+        }
+
+        private static void VerifyCmsCorrectlySigned(byte[] content, byte[] signedData, X509Certificate2 cert, X509Certificate2 counterSignerCert = null)
+        {
+            Assert.False(cert.HasPrivateKey);
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            cms.Decode(signedData);
+            cms.CheckSignature(verifySignatureOnly: true);
+
+            Assert.Equal(1, cms.SignerInfos.Count);
+
+            SignerInfo signer = cms.SignerInfos[0];
+            Assert.Equal(cert, signer.Certificate);
+
+            if (counterSignerCert != null)
+            {
+                Assert.Equal(1, signer.CounterSignerInfos.Count);
+                Assert.Equal(counterSignerCert, signer.CounterSignerInfos[0].Certificate);
+            }
+            else
+            {
+                Assert.Equal(0, signer.CounterSignerInfos.Count);
+            }
+        }
+
+        private static (byte[] signedData, X509Certificate2 pubCert) GetSignedData(byte[] content, Func<(X509Certificate2, AsymmetricAlgorithm)> getCertAndKey, Func<AsymmetricAlgorithm, bool> checkIfSignHashCalled)
+        {
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            (X509Certificate2 cert, AsymmetricAlgorithm key) = getCertAndKey();
+            using (key)
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                if (key is DSA)
+                {
+                    signer.IncludeOption = X509IncludeOption.EndCertOnly;
+                    signer.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                }
+
+                cms.ComputeSignature(signer);
+
+                byte[] ret = cms.Encode();
+                Assert.True(checkIfSignHashCalled(key));
+
+                Assert.True(ByteArrayContains(ret, content));
+
+                return (ret, cert);
+            }
+        }
+
+        private static (byte[] counterSignedData, X509Certificate2 firstSignerCert, X509Certificate2 counterSignerCert) GetCounterSignedData(byte[] content, Func<(X509Certificate2, AsymmetricAlgorithm)> getFirstSignerCertAndKey, Func<(X509Certificate2, AsymmetricAlgorithm)> getCounterSignerCertAndKey, Func<AsymmetricAlgorithm, bool> signedHashCalled, Func<AsymmetricAlgorithm, bool> signedHashCalledForCounterSignerKey)
```

Comment:
I realize this line is too long but no clue what are the guidelines for formatting functions returning tuples